### <a name="change-in-behavior-for-taskwaitall-methods-with-time-out-arguments"></a><span data-ttu-id="f9b38-101">Изменение в поведении методов Task.WaitAll с аргументами времени ожидания</span><span class="sxs-lookup"><span data-stu-id="f9b38-101">Change in behavior for Task.WaitAll methods with time-out arguments</span></span>

|   |   |
|---|---|
|<span data-ttu-id="f9b38-102">Подробные сведения</span><span class="sxs-lookup"><span data-stu-id="f9b38-102">Details</span></span>|<span data-ttu-id="f9b38-103">В .NET 4.5 реализовано согласованное поведение методов Task.WaitAll. В .NET Framework 4 поведение этих методов было непоследовательным.</span><span class="sxs-lookup"><span data-stu-id="f9b38-103">Task.WaitAll behavior was made more consistent in .NET 4.5.In the .NET Framework 4, these methods behaved inconsistently.</span></span> <span data-ttu-id="f9b38-104">По истечении времени ожидания, если одна или несколько задач были завершены или отменены до вызова метода, метод вызывал исключение <xref:System.AggregateException?displayProperty=name>.</span><span class="sxs-lookup"><span data-stu-id="f9b38-104">When the time-out expired, if one or more tasks were completed or canceled before the method call, the method threw an <xref:System.AggregateException?displayProperty=name> exception.</span></span> <span data-ttu-id="f9b38-105">По истечении времени ожидания, если ни одна задача не была завершена или отменена до вызова метода, однако одна или несколько задач входили в эти состояния после вызова метода, метод возвращал значение false.</span><span class="sxs-lookup"><span data-stu-id="f9b38-105">When the time-out expired, if no tasks were completed or canceled before the method call, but one or more tasks entered these states after the method call, the method returned false.</span></span><br/><br/><span data-ttu-id="f9b38-106">В .NET Framework 4.5 эти перегрузки метода теперь возвращают false, если все задачи по-прежнему выполняются после истечения времени ожидания, и вызывают исключение <xref:System.AggregateException?displayProperty=name>, только если входная задача была отменена (независимо от того, была ли она до или после вызова метода), а никакие другие задачи не выполняются.</span><span class="sxs-lookup"><span data-stu-id="f9b38-106">In the .NET Framework 4.5, these method overloads now return false if any tasks are still running when the time-out interval expired, and they throw an <xref:System.AggregateException?displayProperty=name> exception only if an input task was cancelled (regardless of whether it was before or after the method call) and no other tasks are still running.</span></span>|
|<span data-ttu-id="f9b38-107">Предложение</span><span class="sxs-lookup"><span data-stu-id="f9b38-107">Suggestion</span></span>|<span data-ttu-id="f9b38-108">Если исключение <xref:System.AggregateException?displayProperty=name> было перехвачено с точки зрения обнаружения задачи, которая была отменена до вызова метода WaitAll, этот код должен выполнить то же обнаружение с помощью свойства IsCanceled (например, .Any(t =&gt; t.IsCanceled)), так как .NET 4.6 создаст исключение только в том случае, если все ожидаемые задачи завершены до истечения времени ожидания.</span><span class="sxs-lookup"><span data-stu-id="f9b38-108">If an <xref:System.AggregateException?displayProperty=name> was being caught as a means of detecting a task that was cancelled prior to the WaitAll call being invoked, that code should instead do the same detection via the IsCanceled property (for example: .Any(t =&gt; t.IsCanceled)) since .NET 4.6 will only throw in that case if all awaited tasks are completed prior to the timeout.</span></span>|
|<span data-ttu-id="f9b38-109">Область</span><span class="sxs-lookup"><span data-stu-id="f9b38-109">Scope</span></span>|<span data-ttu-id="f9b38-110">Дополнительный номер</span><span class="sxs-lookup"><span data-stu-id="f9b38-110">Minor</span></span>|
|<span data-ttu-id="f9b38-111">Версия</span><span class="sxs-lookup"><span data-stu-id="f9b38-111">Version</span></span>|<span data-ttu-id="f9b38-112">4.5</span><span class="sxs-lookup"><span data-stu-id="f9b38-112">4.5</span></span>|
|<span data-ttu-id="f9b38-113">Тип</span><span class="sxs-lookup"><span data-stu-id="f9b38-113">Type</span></span>|<span data-ttu-id="f9b38-114">Среда выполнения</span><span class="sxs-lookup"><span data-stu-id="f9b38-114">Runtime</span></span>|
|<span data-ttu-id="f9b38-115">Затронутые API</span><span class="sxs-lookup"><span data-stu-id="f9b38-115">Affected APIs</span></span>|<ul><li><xref:System.Threading.Tasks.Task.WaitAll(System.Threading.Tasks.Task[],System.Int32)?displayProperty=nameWithType></li><li><xref:System.Threading.Tasks.Task.WaitAll(System.Threading.Tasks.Task[],System.Int32,System.Threading.CancellationToken)?displayProperty=nameWithType></li><li><xref:System.Threading.Tasks.Task.WaitAll(System.Threading.Tasks.Task[],System.TimeSpan)?displayProperty=nameWithType></li></ul>|

